src/site/_includes/components/Prerender.js

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const {html} = require('common-tags');
+
+/**
+ * Prerender by adding Speculation Rules JSON
+ * Also adds a GA event so can see how many prerender attempts were made
+ * versus prerender navigations.
+ * @param {string} url The url to prerender
+ * @returns {string}
+ */
+module.exports = (url) => {
+  // Only prerender internal URLs
```

Comment:
Note that this logic assumes we never fully write out same-origin URLs.

I'm pretty sure that's true throughout the site, but I'm not sure there's anything enforcing it, so this is relying on us following a convention—and if you're going to do that you could just as easily follow a convention to not prerender cross-origin URLs.

(I.e. I'm not sure how much value this logic is adding.)

Reply:
If you really wanted to check this, a more robust way would be something like this:


```js
const ORIGIN = 'https://web.dev';

// ...

if (new URL(url, ORIGIN).origin !== ORIGIN) {
  return '';
}
```

Reply:
Yeah I thought about that but there is a linting rule enforcing this.

And adding logic to check the domain matches (especially for local dev and staging URLs is more complicated than it's worth IMHO.

Reply:
OK added that code. Thanks.